packages/scrollable-image/src/components/with-waypoints.js

```diff
@@ -36,55 +33,75 @@ const withWaypoints = WrappedComponent => {
       })
     }
 
-    _handleTopBoundaryEnter({ previousPosition, currentPosition }) {
-      if (!this.isEnabled) return
-      if (
-        previousPosition === Waypoint.above &&
-        currentPosition === Waypoint.inside
-      ) {
-        this.setScrollState({
-          isActive: false,
-          verticalDirection: 'up',
-        })
+    _handleTopBoundaryPositionChange(position) {
+      const { previousPosition, currentPosition } = position
+
+      // top boundary enter to viewport
+      if (currentPosition === Waypoint.inside) {
+        if (previousPosition === Waypoint.above) {
+          this.setScrollState({
+            isActive: false,
+            verticalDirection: 'up',
+          })
+        } else if (previousPosition === Waypoint.below) {
+          this.setScrollState({
+            isActive: false,
+            verticalDirection: 'up',
+          })
+        }
+        return
       }
-    }
 
-    _handleTopBoundaryLeave({ previousPosition, currentPosition }) {
-      if (!this.isEnabled) return
-      if (
-        previousPosition === Waypoint.inside &&
-        currentPosition === Waypoint.above
-      ) {
-        this.setScrollState({
-          isActive: true,
-          verticalDirection: 'down',
-        })
+      // top boundary leave from viewport
+      if (previousPosition === Waypoint.inside) {
+        if (currentPosition === Waypoint.above) {
+          this.setScrollState({
+            isActive: true,
+          })
+        } else if (currentPosition === Waypoint.below) {
+          this.setScrollState({
+            isActive: false,
+            verticalDirection: 'up',
+          })
+        }
```

Comment:
So, could we make this `_handleTopBoundaryPositionChange` simpler?
Such as
```
_handleTopBoundaryPositionChange() {
  if (previousPosition === Waypoint.inside && 
    currentPosition === Waypoint.above ) {
    this.setScrollState({
      isActive: true,
    })
    return 
  }
  this.setScrollState({
    isActive: false,
    childrenAligned: 'top', 
  })
}
```

If you think this change is ok,
the same modification could be applied to `_handleBottomBoundaryPositionChange`.

Reply:
Current `_handleTopBoundaryPositionChange` seems cumbersome, but it can clearly state which case it handle and how does it handle (use `this.setScrollState`). From this perspective, sample code snippet  in this comment can not provide the certain extent of clarity as current solution.

Moreover, there are more cases should be handled in code snippet above, such as `previousPosition === Waypoint.above` and `currentPosition === Waypoint.below`, which increases the complexity of conditions and  readability.

If current conditions are not clear enough, I could append more comments to them.

Reply:
> Moreover, there are more cases should be handled in code snippet above, such as previousPosition === Waypoint.above and currentPosition === Waypoint.below, which increases the complexity of conditions and become less readable.

`previousPosition === Waypoint.above` 且 `currentPosition === Waypoint.below` 這個情況就我們的例子應該不會出現。
因為 Waypoint 偵測的 viewport 是 100vh（我們討論的前提是不加 `bottomOffset="99%"`）。因此，如果 `currentPosition === Waypoint.below` 的話，那`previousPosition` 應該是 `Waypoint.inside`。

Reply:
Updated.